pkg/controller/certificates/trigger/policies/gatherer_test.go

```diff
@@ -0,0 +1,336 @@
+				assert.Error(t, gotErr)
+				assert.EqualError(t, gotErr, test.wantErr)
+				return
+			}
+
+			require.NoError(t, gotErr)
+			assert.Equal(t, test.wantRequest, got.CurrentRevisionRequest)
+			assert.Equal(t, test.wantSecret, got.Secret)
+			assert.Equal(t, test.givenCert, got.Certificate, "input cert should always be equal to returned cert")
+		})
+	}
+}
+
+// Creates a mock CertificateRequestLister.
+//
+// We want to use a mock instead of a fake here: the mock makes sure that
+// (1) the lister.CertificateRequests(namespace) has been called with the
+// correct expected namespace and (2) lister.CertificateRequests(namespace)
+// has been called exactly once, which makes sure (1) was checked.
+func mockCertificateRequests(expNamespace string, innerLister func(t *testing.T) *listers.FakeCertificateRequestNamespaceLister) func(*testing.T) *listers.FakeCertificateRequestLister {
+	return func(t *testing.T) *listers.FakeCertificateRequestLister {
+		f := expectCalled(t, 1)
+		return listers.
+			NewFakeCertificateRequestLister().
+			WithCertificateRequests(func(namespace string) cmlist.CertificateRequestNamespaceLister {
+				f()
+				assert.Equal(t, expNamespace, namespace)
+				return innerLister(t)
+			})
+	}
+}
+
+// The returned function f is expected to be called expectedCount times.
+// Returns a friendly error with the location where f was created to help
+// the developer figure out where the error came from.
+func expectCalled(t *testing.T, expectedCount uint32) (f func()) {
+	// Debugging purposes: allows us to display where this function was
+	// called. The reason we need to to this is because t.Cleanup's stack
+	// does not contain the origin function, which makes it really hard to
+	// know where the assertion failed.
+	stack := grandparentStack(2)
+
+	gotCount := uint32(0)
+	t.Cleanup(func() {
+		assert.Equal(t, int(expectedCount), int(atomic.LoadUint32(&gotCount)), "this func was expected to be called %d times but was called %d times. Stack:\n\t%s",
+			expectedCount,
+			gotCount,
+			strings.Join(stack, "\n\t"),
+		)
+	})
+	return func() {
+		atomic.AddUint32(&gotCount, 1)
+	}
+}
+
+func expectNeverCalled() func(t *testing.T) *listers.FakeCertificateRequestLister {
+	return func(t *testing.T) *listers.FakeCertificateRequestLister {
+		f := expectCalled(t, 0)
+		return listers.NewFakeCertificateRequestLister().WithCertificateRequests(func(namespace string) cmlist.CertificateRequestNamespaceLister {
+			f()
+			return nil
+		})
+	}
+}
+
+// Returns the stack of the callers of the form "file.go:93". The N first
+// items are skipped. The stack stops as soon as it meets a function in
+// testing.go since we are not interested by those.
+func grandparentStack(skip int) []string {
+	var stack []string
+	for i := skip; ; i++ {
+		_, file, line, ok := runtime.Caller(i)
+		file = filepath.Base(file)
+		if file == "testing.go" || !ok {
+			break
+		}
+		stack = append(stack, fmt.Sprintf("%s:%d", file, line))
+	}
+	return stack
+}
```

Comment:
This seems unnecessary

Reply:
That said, I can always remove this `whereAmI()` thing if you think it is pointless

Reply:
→ moved to using the fake clientset instead